[MRG] ENH Add parameter to sklearn.manifold.TSNE to transition to squaring non-Euclidean distance metrics by default (#12401)

joshuacwnewton opened the pull request:
#### Reference Issues/PRs

Fixes #12401
Supersedes and closes #12419

The plan for deprecating legacy squaring behaviour is summarized in [this comment.](https://github.com/scikit-learn/scikit-learn/issues/12401#issuecomment-652893433)

#### What does this implement/fix? Explain your changes.

 * Adds a `square_distances` parameter to TSNE. (`default='legacy'`)
     * `True` enables squaring for all distances metrics
     * `'legacy'` preserves existing behaviour, where distances will only be squared when `metric='euclidean'`. Raises a `FutureWarning` when a non-Euclidean metric is provided to alert the deprecation of legacy behavior. Setting `True` suppresses this warning.
 * Adds a parameter description to the docstring of TSNE
 * Adds tests to check `square_distances` and check that `FutureWarning` is raised properly.
 * Modifies existing tests to include `square_distances=True` to suppress FutureWarnings.

#### Any other comments?

* ~~I was uncertain about the usage of `pytest.mark.parametrize` in my tests. Some existing tests use it, while others iterate through lists of parameters. Style suggestions are appreciated.~~ Answered in feedback.
* ~~I was uncertain about the descriptions for my tests. Some use docstrings while others use a simple comment. Style suggestions are appreciated.~~ Answered in feedback.
* I was uncertain about proposing a timeline for changing the default value and/or deprecating `'legacy'` behaviour. Discussion in the issue (#12401) has become outdated, so I chose version numbers based on the deprecation section of the Contributing docs.

joshuacwnewton committed:
Add new metric squaring behavior to TSNE

joshuacwnewton committed:
Fix truth value check for string value cases

joshuacwnewton committed:
Remove one non-neg check when square_distance=True

joshuacwnewton committed:
Update non_positive test to ensure no squaring

joshuacwnewton committed:
Revert non-neg check for precomputed distances

I presumed that adding square_distance could allow for negative
precomputed distances in some cases (square_distance=True). However,
`knn.fit()` does its own non-neg check, which forces precomputed
distances to be positive regardless of the new square_distance
parameter.

joshuacwnewton committed:
Add tests for changes to square_distance

Tests cover the following areas:
    * Invalid square_distance parameter selection
    * TSNE functioning with metric=Euclidean and square_distance=legacy
    * TSNE functioning with square_distance=True/False

joshuacwnewton committed:
Add docstring entry for squared_distance parameter

joshuacwnewton committed:
Add test for square_distance FutureWarning cases

joshuacwnewton committed:
Revert change to non-neg check print statement

joshuacwnewton commented:
It looks like adding a `FutureWarning` caused other TSNE tests to fail. I'm going to add `@ignore_warnings(category=FutureWarning)` to the relevant tests, but I'm unsure if that's the proper way to go here. Happy to change on further feedback.

A related question: should I be running my local tests to fail on Warnings, then? I must have not been using the right `pytest` flags, as I didn't catch this before opening a PR.

joshuacwnewton committed:
Ignore `FutureWarning` in relevant TSNE tests

joshuacwnewton committed:
Remove parameterization to speed up slow TSNE test

joshuacwnewton committed:
Restore n_jobs param to pairwise_distances call

joshuacwnewton commented:
Pinging @flosincaptie. I think this is ready to go from [WIP] -> [MRG]. :)

flosincapite commented on sklearn/manifold/_t_sne.py:
```diff
@@ -597,6 +598,14 @@ class TSNE(BaseEstimator):
 
         .. versionadded:: 0.22
 
+    square_distance : {True, False, or 'legacy'}, default='legacy'
```
The standard way to do something like this is an [enum](https://docs.python.org/3/library/enum.html). 

Although Python allows arguments to be of arbitrary types, it's usually cleaner to accept only one type (or "interface," e.g. "any iterable," i.e. "anything that implements `__iter__` and `__next__`").

thomasjpfan replied:
We traditionally prefer not to have an enum here because the enum would have to be scoped to this specific parameter for this specific estimator. From a CS point of view, it would be cleaner to have a single type. But for our users with non-CS backgrounds, reducing the number of abstractions, especially in the public interface, makes sklearn easy to use.

joshuacwnewton replied:
If I understand correctly, scitkit-learn requires input validation to be done in `.fit()` specifically, rather than enforcing it in the constructor. I'm trying to find the source where I read this... [this section](https://scikit-learn.org/dev/developers/develop.html#instantiation) is the closest I've come, but I don't think it's exactly it.

thomasjpfan replied:
Yes, validation is done in `fit`.

joshuacwnewton replied:
Oops, sorry, I hadn't refreshed the page. I missed all of your comments when I made mine. So sorry, @thomasjpfan.

flosincapite commented on sklearn/manifold/_t_sne.py:
```diff
@@ -714,16 +733,19 @@ def _fit(self, X, skip_num_points=0):
-                else:
-                    distances = pairwise_distances(X, metric=self.metric,
-                                                   n_jobs=self.n_jobs)
+                distances = pairwise_distances(X, metric=self.metric,
```
Why the switch here?

joshuacwnewton replied:
By perform squaring [a little further down](https://github.com/scikit-learn/scikit-learn/pull/17662/files/88340e6786e1a8be09a2bb8ea9c22eeadd4dd6ac#diff-e9ec8beb584b0141adb0b124e88fbd40R744) instead, it allowed me to combine the legacy case with the general case, which I felt was a bit cleaner.

jeremiedbb replied:
pairwise_distances with euclidean_distances is know to be slower with n_jobs > 1 so we should keep the branching.

Also it's better to pass `squared` for euclidean since it first computes the squared distance matrix. Otherwise you'll compute the squared distance matrix, then take the square root and then square it again.

joshuacwnewton replied:
Thank you for the insight. I'll add a comment to make this reasoning clearer. :)

flosincapite commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -277,6 +278,7 @@ def test_fit_csr_matrix(method):
                     1.0, rtol=1.1e-1)
 
 
+@ignore_warnings(category=FutureWarning)
```
Why not update the tests to use `square_distance=True`?

That way, when the API changes, these tests will already be correct--no need to remove a bunch of `ignore_warnings` pragmas :).

You can also add additional happy-path tests here to exercise the code paths where `square_distance` is `False` or `'legacy'`.

flosincapite replied:
Ignore the last paragraph; I see you already added those tests.

joshuacwnewton replied:
My thinking was that I wanted to make minimal changes to the bodies of the tests themselves in this PR. 

Setting square_distance=True would change the behavior of those tests to the future functionality. I wanted to delay that change until the default is truly updated in 0.26.

jeremiedbb replied:
I agree that we need tests for true, false and legacy. However for the tests that only check that proper errors are raised for wrong parameters we can safely switch to the new behavior.

jeremiedbb replied:
Also it would be good to make sure that all tests still pass with squared_distances = True

joshuacwnewton replied:
Thank you for the input! For now, I've made this change by parametrizing `square_distance` to check `True` and `False` across relevant tests. But, as was mentioned elsewhere, I have mixed feelings about the `square_distance=False` behavior in general.

flosincapite commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -879,6 +897,71 @@ def test_tsne_with_different_distance_metrics():
+    assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
+
+
+def test_tsne_with_different_square_distances():
```
Ignore what I said above--you already added these happy-path tests.

flosincapite reviewed (changes requested):
Looks great! There are some comments, but only because I was _extremely_ picky :)

joshuacwnewton commented:
Thanks a ton for taking the time to review, @flosincapite! :D

joshuacwnewton committed:
Fix calls to pairwise_distances for Euclidean case

Euclidean handles squaring differently than other metrics. Euclidean
is also slower for n_jobs > 1. So, the same call to pairwise_distances
can't be used for every metric -- they have to be separated.

joshuacwnewton committed:
Write test to demonstrate squaring differences

joshuacwnewton commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -921,6 +921,27 @@ def test_tsne_with_legacy_euclidean_squaring(method):
     assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
 
 
+@pytest.mark.parametrize('method', ['exact', 'barnes_hut'])
+@ignore_warnings(category=FutureWarning)
+def test_tsne_with_different_squaring_methods(method):
+    """Isolate failing behavior in
+    `test_tsne_with_legacy_euclidean_squaring.`"""
+    random_state = check_random_state(0)
+    n_components_original = 3
+    n_components_embedding = 2
+    X = random_state.randn(50, n_components_original).astype(np.float32)
+    X_precomputed_1 = pairwise_distances(X, metric='euclidean') ** 2
+    X_precomputed_2 = pairwise_distances(X, metric='euclidean', squared=True)
+    X_transformed_tsne_precomputed_1 = TSNE(
+        metric='precomputed', n_components=n_components_embedding,
+        random_state=0, method=method).fit_transform(X_precomputed_1)
+    X_transformed_tsne_precomputed_2 = TSNE(
+        metric='precomputed', n_components=n_components_embedding,
+        random_state=0, method=method).fit_transform(X_precomputed_2)
+    assert_array_equal(X_transformed_tsne_precomputed_1,
+                       X_transformed_tsne_precomputed_2)
+
+
```
While making the requested change in 08d7b79793f795eb8c370837563f72a899e9ce30, I discovered a separate issue: using different squaring methods results in different TSNE results, i.e. the test here fails.

The reason the difference is relevant is because of `'barnes_hut'` method. 
* If distances are not precomputed, it uses `**=2` to square [here](https://github.com/scikit-learn/scikit-learn/blob/5edf203d6ee8a168c127b7bff7410ac83b497a30/sklearn/manifold/_t_sne.py#L809). 
* If distances are precomputed, `squared=True` could be used, e.g. in [this other test](https://github.com/scikit-learn/scikit-learn/pull/17662/files#diff-bf12bea42d63e12299a1eee54411bcc6R916).

These two cases produce different results. I believe this behavior was present before any of the changes in this PR.

joshuacwnewton replied:
To be clear: [`test_tsne_with_legacy_euclidean_squaring`](https://github.com/scikit-learn/scikit-learn/pull/17662/files#diff-bf12bea42d63e12299a1eee54411bcc6R902) currently passes when `exact` is used, and fails when `barnes_hut` is used. The above behavior is why.

joshuacwnewton replied:
I explained this in an unclear way. I'm going to try and reword my thoughts.

jeremiedbb replied:
```
X_precomputed_1 = pairwise_distances(X, metric='euclidean') ** 2
X_precomputed_2 = pairwise_distances(X, metric='euclidean', squared=True)
```

Although mathematically the same, they might slightly differ in floating arithmetics. Then we can expect the results of t-sne to be slightly different too. Usually we use `assert_allclose` to compare arrays of floats. Since you're using float32, you can change the rtol to 1e-6 or even 1e-5.

joshuacwnewton replied:
Ah, sorry, it might have been better to also include the output of the failing tests.

```
E       Mismatched elements: 100 / 100 (100%)
E       Max absolute difference: 182.95767
E       Max relative difference: 35.374535
E        x: array([[-36.36463 , -60.091194],
E              [ 35.217983, -77.36719 ],
E              [-23.366423,  -1.641921],...
E        y: array([[ -83.61592 ,  -40.051804],
E              [ -12.799014,  105.59048 ],
E              [  -2.169685,  -25.965828],...
```

A small change in the distances (e.g. due to float arithmetic) seems to have caused a large difference in the TSNE output. So, I'm not sure `assert_allclose` would help here, but I could be mistaken.

jnothman replied:
I'm not entirely surprised that tSNE is unstable. I can confirm that the differences are small:

absolute diff between `X_precomputed_1` and `X_precomputed_2`:
![absdiff](https://user-images.githubusercontent.com/78827/85648625-7cc05680-b6e4-11ea-9e51-7b4e48f36393.png)

Relative diff:
![reldiff](https://user-images.githubusercontent.com/78827/85648632-81850a80-b6e4-11ea-9cae-bca5b66af2b9.png)

> Although mathematically the same, they might slightly differ in floating arithmetics.

More explicitly, in `X_precomputed_1 = pairwise_distances(X, metric='euclidean') ** 2` the computation loses precision when taking the square root in euclidean_distances.

joshuacwnewton replied:
> I'm not entirely surprised that tSNE is unstable. I can confirm that the differences are small:

Thank you for taking the time to verify this, and to share the plots you've created. :)

jnothman replied:
Are there perhaps parameters we can set to make the tSNE optimisation more stable? I would have still expected to land up with the same nearest neighbors graph, but slightly different probability distribution.

joshuacwnewton replied:
Quick question: Would it be better to have this chat in a separate issue, just to open up the conversation a bit? 

While distance squaring was what started the conversation, from my understanding any small differences could trigger this behaviour, so it might be a bit off-topic from this PR.

jnothman replied:
I'm not sure if the instability of tSNE is an issue as such... But it's an issue with regards to how to test this change!

joshuacwnewton replied:
Aha, that's good to know! I had assumed instability was a general issue, so thank you for clarifying. :)

A quick/dirty way to fix the `test_tsne_with_legacy_euclidean_squaring` test would be to replace [this line](https://github.com/scikit-learn/scikit-learn/pull/17662/files#diff-bf12bea42d63e12299a1eee54411bcc6R916) with:

```python
if method == 'barnes_hut':
    X_precomputed = pairwise_distances(X, metric='euclidean') ** 2
else:
    X_precomputed = pairwise_distances(X, metric='euclidean', squared=True)
```
That way the squaring method matches up with what's used internally in `TSNE`. But, if I understand correctly, you're asking if there's a way to not have to resort to this?

joshuacwnewton replied:
Going out on a limb here, but the lack of stability could also have something do with with the use of `randn` to generate sample pairwise distances. Without much structure in our sample data, t-SNE might not find meaningful topology in the data, as is shown in [this tutorial](https://scikit-learn.org/stable/auto_examples/manifold/plot_t_sne_perplexity.html#sphx-glr-auto-examples-manifold-plot-t-sne-perplexity-py).

I wonder if it would be better to create/reuse a small toy dataset for TSNE tests that has structure. This could better reflect the intended use cases for t-SNE.

joshuacwnewton replied:
Yep, the test pass with `make_blobs`. I''ll update accordingly.

joshuacwnewton replied:
Yes, it does seem like `make_blobs` fixed the stability issues. :)

jnothman replied:
That makes a lot of sense. Good idea :)

joshuacwnewton replied:
Thank you for your help! Your questions pointed me in the right direction. :)

joshuacwnewton commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -879,6 +897,93 @@ def test_tsne_with_different_distance_metrics():
         assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
 
 
+@pytest.mark.parametrize('method', ['exact', 'barnes_hut'])
+@ignore_warnings(category=FutureWarning)
+def test_tsne_with_legacy_euclidean_squaring(method):
+    """Make sure that TSNE preserves legacy Euclidean squaring behavior
+
+    Note: If 'legacy' is deprecated when the default for square_distance
+    is set to True, then this test can be removed, and 'euclidean' can
+    be included in `test_tsne_with_different_distance_metrics`."""
+    random_state = check_random_state(0)
+    n_components_original = 3
+    n_components_embedding = 2
+    X = random_state.randn(50, n_components_original).astype(np.float32)
+    X_transformed_tsne = TSNE(
+        metric='euclidean', n_components=n_components_embedding,
+        random_state=0, square_distance="legacy",
+        method=method).fit_transform(X)
+    X_precomputed = pairwise_distances(X, metric='euclidean', squared=True)
+    X_transformed_tsne_precomputed = TSNE(
+        metric='precomputed', n_components=n_components_embedding,
+        random_state=0, square_distance="legacy",
+        method=method).fit_transform(X_precomputed)
+    assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
```
I believe I've uncovered an existing issue with this test. This test passes with `'exact'`, but fails with `'barnes_hut'`.

I think the reason why the `'barnes_hut'` test fails is because it uses `**=2` to square the distances, [here](https://github.com/scikit-learn/scikit-learn/blob/5edf203d6ee8a168c127b7bff7410ac83b497a30/sklearn/manifold/_t_sne.py#L809). But, the precomputed distance in this test uses `squared=True`.

To verify this is the case, I've written [a second test](https://github.com/scikit-learn/scikit-learn/pull/17662/files#diff-bf12bea42d63e12299a1eee54411bcc6R945) which intentionally fails. It shows that TSNE produces different results depending on which squaring method is used (`**=2` or `squared=True`).

joshuacwnewton committed:
Replace random test data with clustered data

t-SNE relies on structure in input data. When random data is used,
t-SNE becomes very sensitive to slight changes in the input. So,
make_blobs is used to create data with structure instead.

joshuacwnewton committed:
Remove test for comparing squaring methods

No longer necessary now that source of instability has been identified.

joshuacwnewton committed:
Refactor tests to be more readable

joshuacwnewton committed:
Parametrize square_dist test to improve coverage

joshuacwnewton committed:
Remove whitespace from existing test

I thought I had written this test, but it wasn't mine! So, this change
shouldn't have happened in the first place.

joshuacwnewton committed:
Clarify comment re: make_blobs and instability

jeremiedbb commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -879,6 +896,84 @@ def test_tsne_with_different_distance_metrics():
+        random_state=0, square_distance="legacy",
+        method=method).fit_transform(X_precomputed)
+
+    assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
```
assert_allclose is preferred when comparing float arrays

```suggestion
    assert_allclose(X_transformed_tsne, X_transformed_tsne_precomputed)
```

joshuacwnewton replied:
Thank you, this makes sense. :)

jeremiedbb commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -879,6 +896,84 @@ def test_tsne_with_different_distance_metrics():
+                metric='precomputed', n_components=n_components_embedding,
+                square_distance=square_distance, method=method,
+                random_state=0).fit_transform(X_precomputed)
+            assert_array_equal(X_transformed_tsne,
+                               X_transformed_tsne_precomputed)
```
```suggestion
            assert_allclose(X_transformed_tsne, X_transformed_tsne_precomputed)
```

jeremiedbb commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -879,6 +896,84 @@ def test_tsne_with_different_distance_metrics():
+
+    for metric in ['euclidean', 'manhattan']:
+        for square_distance in [True, False, 'legacy', '']:
+            if square_distance == '':
```
This is the same as letting the default value of square_distance. You can remove this case

jeremiedbb commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -879,6 +896,84 @@ def test_tsne_with_different_distance_metrics():
+
+    X = random_state.randn(5, 2)
+
+    for metric in ['euclidean', 'manhattan']:
+        for square_distance in [True, False, 'legacy', '']:
```
You can parametrize this

jeremiedbb reviewed (commented):
I made a few comments. I don't think we should enable the possibility to have `metric="euclidean"` and `square_distance=False`.

joshuacwnewton commented on sklearn/manifold/_t_sne.py:
```diff
@@ -714,16 +733,28 @@ def _fit(self, X, skip_num_points=0):
-                if np.any(distances < 0):
-                    raise ValueError("All distances should be positive, the "
-                                     "metric given is not correct")
+            if self.metric != 'euclidean' and self.square_distance is True:
```
I believe this change would cause non-Euclidean distances to be squared when `square_distance=='legacy'`, because the truth value for non-empty strings is considered True.

jeremiedbb replied:
right I forgot it was not just a boolean

joshuacwnewton replied:
No worries! :)

joshuacwnewton commented on sklearn/manifold/_t_sne.py:
```diff
@@ -714,16 +733,28 @@ def _fit(self, X, skip_num_points=0):
+                        #   ^2 -> sqrt() -> ^2
+                        distances = pairwise_distances(X, metric=self.metric,
+                                                       squared=True)
+                    else:
+                        distances = pairwise_distances(X, metric=self.metric)
```
That's a good point. Yes... currently, with legacy behavior, it's not even possible to have non-squared Euclidean distances. Thanks for pointing that out. :)

joshuacwnewton replied:
It does feel a bit strange to disallow `square_distance=False` for one metric, but allow it for other metrics. 

Is it clear to the user why they would ever even want to set `square_distance=False`?

jeremiedbb replied:
> Is it clear to the user why they would ever even want to set square_distance=False

I don't think so indeed

joshuacwnewton replied:
I have brought this up in the issue [here](https://github.com/scikit-learn/scikit-learn/issues/12401#issuecomment-650288735). I am curious to hear thoughts on this. :)

joshuacwnewton replied:
Based on discussion in #12401, I think I'm going to update the PR to remove `'False'` as an option, which should clear up this problem.

joshuacwnewton committed:
Shorten Euclidean squaring comment

joshuacwnewton committed:
Replace assert_all_equal with allclose in tests

joshuacwnewton committed:
Replace ignore(FutureWarning) with parametrization

With the addition of `square_distance`, existing tests now raise a
FutureWarning. This could be ignored using ignore_warnings, but it's
better to ensure existing tests pass regardless of whether
`square_distance` is set to True or False.

joshuacwnewton committed:
Remove default case from FutureWarning test

joshuacwnewton committed:
Parametrize square_distance_futurewarning test

joshuacwnewton committed:
Fix linting issue (indentation)

joshuacwnewton committed:
Shorten test names to pass linting (line length)

This required me to change the names of existing tests. I believe this
is discouraged, but I'm unsure if this can be achieved another way.

joshuacwnewton committed:
Change square_distance to {True, 'legacy', 'warn'}

In discussion in #12401, it was decided that False is not necessary, so
this behavior is removed. 'warn' is added to allow 'legacy' behavior
to be explicitly set without warning. Tests are updated accordingly.

joshuacwnewton committed:
Restore long test names

Two existing test names were shortened when parametrization was added.
This parametrization is no longer necessary, so restore the old names.

joshuacwnewton committed:
Remove parametrization in neighbours TSNE test

joshuacwnewton committed:
Update FutureWarning message to reflect changes

False was removed, and 'legacy' can now suppress the warning.

joshuacwnewton committed:
Merge remote-tracking branch 'upstream/master' into 12401-add-TSNE-squaring

jeremiedbb commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -872,13 +884,64 @@ def test_tsne_with_different_distance_metrics():
+@pytest.mark.parametrize('method', ['exact', 'barnes_hut'])
+@pytest.mark.parametrize('metric', ['euclidean', 'manhattan'])
+@pytest.mark.parametrize('square_distance', [True, 'legacy', 'warn'])
+@ignore_warnings(category=FutureWarning)
```
Please add a comment saying
"FIXME remove when square_distance=True becomes the default in 0.26"

jeremiedbb reviewed (commented):
Small comments. Otherwise looks good @joshuacwnewton

Also please add a what's new entry

joshuacwnewton commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -371,7 +373,8 @@ def test_non_positive_computed_distances():
         return -1
 
-    tsne = TSNE(metric=metric, method='exact')
+    # For square_distance=True, no ValueError will be raised
+    tsne = TSNE(metric=metric, method='exact', square_distance='legacy')
```
Aha, good thinking. Thank you. :)

joshuacwnewton committed:
Check neg distances before squaring, update test

Negative distances should be caught, and an error should be thrown,
even if squaring would remove the negative. Updates the relevant test
to ensure squaring doesn't affect the test.

joshuacwnewton committed:
Remove 'or' from {True, 'legacy', 'warn'}

joshuacwnewton committed:
Add "FIXME" comment for FutureWarning test

joshuacwnewton committed:
Merge remote-tracking branch 'upstream/master' into 12401-add-TSNE-squaring

joshuacwnewton committed:
Add "What's New" entry for square_distance

joshuacwnewton committed:
Make '/" usage consistent for `metric`

joshuacwnewton commented:
Thanks much for the reviews @jeremiedbb. :)

jeremiedbb commented on sklearn/manifold/_t_sne.py:
```diff
@@ -597,6 +598,14 @@ class TSNE(BaseEstimator):
 
         .. versionadded:: 0.22
 
+    square_distance : {True, 'legacy', 'warn'}, default='warn'
```
Forgot that in my last review. We don't put 'warn' in the docstrings.

```suggestion
    square_distance : {True, 'legacy'}, default='legacy'
```

joshuacwnewton replied:
Thanks for the additional comments! :)

jeremiedbb commented on sklearn/manifold/_t_sne.py:
```diff
@@ -597,6 +598,14 @@ class TSNE(BaseEstimator):
+        The default value for this parameter will change to ``True`` in
+        version 0.26.
+
+        .. versionadded:: 0.24
```
Maybe add a small explanation here like 
```
.. versionadded:: 0.24
``square_distance`` is added in 0.24 for back-compatibility purpose.
```

joshuacwnewton committed:
Remove 'warn' from TSNE docstring

joshuacwnewton committed:
Explain deprecation in docstring/"What's New"

joshuacwnewton committed:
Update TSNE to deprecate 'legacy' by 0.26

* Update warning to announce deprecation
* Warn when 'legacy' is set.
* Because 'legacy' now functions exactly the same as 'warn', remove
'warn' as a setting for 'square_distance'.
* Update tests to reflect changes.
* Update docstring to reflect changes.

jeremiedbb commented on sklearn/manifold/_t_sne.py:
```diff
@@ -597,6 +598,16 @@ class TSNE(BaseEstimator):
+    square_distance : {True, 'legacy'}, default='legacy'
+        Whether distance values are squared before t-SNE is run. ``'legacy'``
+        means that distances are squared only when ``metric="euclidean"``.
+        In version 0.26, ``'legacy'`` will be removed as a setting, and the
```
That should go in a `.. deprecated::` section, with a message like "`square_distance='legacy'` is deprecated in 0.24 and will not be possible in 0.26. The default value for this parameter will change to `True`".

joshuacwnewton replied:
Ah, of course, thank you! :)

jeremiedbb commented on sklearn/manifold/_t_sne.py:
```diff
@@ -666,6 +678,16 @@ def _fit(self, X, skip_num_points=0):
+                           "compatibility purposes. However, 'legacy' will "
+                           "be removed in 0.26, at which point all distance "
+                           "metrics will be squared by default. Set "
+                           "'square_distance' to True to silence this "
+                           "warning."), FutureWarning)
```
I would remove the last sentence because setting to True changes the behavior in addition to silence the warning.

jeremiedbb reviewed (approved):
LGTM,  thanks @joshuacwnewton !

joshuacwnewton committed:
Update docstring to use proper deprecation format

joshuacwnewton committed:
Remove last sentence in FutureWarning

joshuacwnewton committed:
Merge upstream/master, fix conflict in What's New

Another enhancement was added to the sklearn.manifold section, causing
a merge conflict.

NicolasHug commented on doc/whats_new/v0.24.rst:
```diff
@@ -158,9 +158,16 @@ Changelog
   :pr:`16289` by :user:`Masashi Kishimoto <kishimoto-banana>` and
   :user:`Olivier Grisel <ogrisel>`.
 
+:mod:`sklearn.manifold`
```
please move this down so that it does not include the isotonic entry below

joshuacwnewton replied:
Ah! Of course, my apologies.

NicolasHug commented on sklearn/manifold/_t_sne.py:
```diff
@@ -597,6 +598,19 @@ class TSNE(BaseEstimator):
         .. versionadded:: 0.22
 
+    square_distance : {True, 'legacy'}, default='legacy'
+        Whether distance values are squared before t-SNE is run. ``'legacy'``
```
It's unclear from this formulation whether we mean "the distances will be squared internally" or "the distances are expected to be squared"

I'd suggest "Whether to square distance values before..."

joshuacwnewton replied:
Ah, very good point. I appreciate the suggestion. :)

NicolasHug commented on sklearn/manifold/_t_sne.py:
```diff
@@ -666,6 +681,15 @@ def _fit(self, X, skip_num_points=0):
             raise ValueError("'method' must be 'barnes_hut' or 'exact'")
         if self.angle < 0.0 or self.angle > 1.0:
             raise ValueError("'angle' must be between 0.0 - 1.0")
+        if self.square_distance not in [True, 'legacy']:
```
Don't we also want to allow for `False`?

Otherwise, when we change the default from legacy to True, True will be the only allowed value for `square_distance`.

CC @jeremiedbb

joshuacwnewton replied:
Thank you much for your review, here and elsewhere. :)

Just for reference, the relevant comment by @jeremiedbb for the current deprecation plan is [this one](https://github.com/scikit-learn/scikit-learn/issues/12401#issuecomment-652893433).

NicolasHug replied:
Thank for the ref. Indeed we don't need to allow for False if don't plan on supporting this case in the future.

I don't think that having `0.26: square_distance={True, 'warn'}. square_distance deprecated` will work however: at this point, we will want to deprecate `square_distance=True` and let users leave it to its default (e.g. None) so that their code doesn't break when we completely remove the parameter. But that's an issue for 0.26 and not for this PR.

joshuacwnewton replied:
Thank you for the input. I've added a `# TODO` referencing this, at the very least.

NicolasHug commented on sklearn/manifold/_t_sne.py:
```diff
@@ -715,15 +739,23 @@ def _fit(self, X, skip_num_points=0):
                     distances = pairwise_distances(X, metric=self.metric,
                                                    squared=True)
                 else:
+                    # Euclidean has separate call as it's slower for n_jobs>1
```
that's not Euclidean, is it?

joshuacwnewton replied:
Ah, yes, my placement of this comment made the message unclear. It was meant to suggest something along the lines of 
```
# The Euclidean case is separate from this call (see above) as it's slower for n_jobs>1
```
Thank you for pointing this out! I'll change it to be more clear. :)

NicolasHug commented on sklearn/manifold/_t_sne.py:
```diff
@@ -715,15 +739,23 @@ def _fit(self, X, skip_num_points=0):
+                raise ValueError("All distances should be positive, the "
+                                 "metric given is not correct")
+
+            if self.metric != "euclidean" and self.square_distance is True:
```
Does this also work in interpreters that aren't CPython? I thought that True being a singleton was a CPython implementation detail

joshuacwnewton replied:
This might not be thorough enough, but a quick search yields:
* IronPython: https://ironpython-test.readthedocs.io/en/latest/library/constants.html
* Jython: https://www.jython.org/jython-old-sites/docs/library/stdtypes.html#boolean-values
* PyPy: I couldn't find reference to bools in https://doc.pypy.org/en/latest/cpython_differences.html#object-identity-of-primitive-values-is-and-id, but I did attempt to quickly test it here: https://pypyjs.org/ See below:

![Screenshot from 2020-07-10 16-17-05](https://user-images.githubusercontent.com/16181459/87210097-37497f00-c2c9-11ea-8239-dbe5a7895b87.png)

NicolasHug commented on sklearn/manifold/_t_sne.py:
```diff
@@ -597,6 +598,19 @@ class TSNE(BaseEstimator):
 
         .. versionadded:: 0.22
 
+    square_distance : {True, 'legacy'}, default='legacy'
```
nit: should this be `square_distances`?

joshuacwnewton replied:
That does seem clearer, yes. :)

NicolasHug commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -872,13 +884,64 @@ def test_tsne_with_different_distance_metrics():
+@pytest.mark.parametrize('square_distance', [True, 'legacy'])
+@ignore_warnings(category=FutureWarning)
+def test_tsne_with_different_square_distances(method, square_distance, metric):
+    """Make sure that TSNE works for different square_distance settings"""
```
Nit: we use comments instead of docstrings for tests

```suggestion
	# Make sure that TSNE works for different square_distance settings
```

joshuacwnewton replied:
I was unsure about this myself, so thank you for the clarification.

NicolasHug reviewed (commented):
thanks @joshuacwnewton, made a first pass, mostly looks good but I'm not quite sure about the deprecation plan

joshuacwnewton committed:
Replace test docstrings with comments

Only new tests were changed. Other existing tests use docstrings, but
were left unchanged.

joshuacwnewton committed:
Fix "What's New" manifold/isotonic header

I accidentally grouped an `sklearn.isotonic` enhancement in
`sklearn.manifold` when fixing a merge conflict.

joshuacwnewton committed:
Clarify wording of `square_distance` description

For previous wording, it was unclear what should be doing the squaring
(TSNE or the user). New wording makes it explicit that TSNE is doing
the squaring.

joshuacwnewton committed:
Clarify meaning of "Euclidean n_jobs>1" comment

joshuacwnewton committed:
Change `square_distance` to `square_distances`

joshuacwnewton committed:
Add TODO note re: deprecation of square_distances

Some implementation details will need to be hashed out in 0.26, so
add reminder.

joshuacwnewton committed:
Fix line length linting issues

`square_distance` -> `square_distances` pushed some lines > 79.

NicolasHug commented on doc/whats_new/v0.24.rst:
```diff
@@ -161,6 +161,13 @@ Changelog
+:mod:`sklearn.manifold`
+.......................
+
+- |Enhancement| Add `square_distance` parameter to :class:`manifold.TSNE`,
```
```suggestion
- |Enhancement| Add `square_distances` parameter to :class:`manifold.TSNE`,
```

joshuacwnewton replied:
Good catch! :)

NicolasHug commented on doc/whats_new/v0.24.rst:
```diff
@@ -161,6 +161,13 @@ Changelog
+.......................
+
+- |Enhancement| Add `square_distance` parameter to :class:`manifold.TSNE`,
+  which provides backward compatibility during deprecation of legacy squaring
+  behavior. :pr:`17662` by :user:`Joshua Newton <joshuacwnewton>`.
```
Let's indicate the plan right away and add:

> Distances will be squared by default in X, and this parameter will be removed in Y.

NicolasHug commented on sklearn/manifold/_t_sne.py:
```diff
@@ -666,6 +682,15 @@ def _fit(self, X, skip_num_points=0):
+                           "'legacy' is provided as a setting for backward "
+                           "compatibility purposes. However, 'legacy' will "
+                           "be removed in 0.26, at which point all distance "
+                           "metrics will be squared by default."),
```
let's add

> Set square_distances=True to avoid this warning. Note that this parameter will be removed in Y.

joshuacwnewton replied:
Hmm. Do you have any thoughts on https://github.com/scikit-learn/scikit-learn/pull/17662#discussion_r450137420?

NicolasHug replied:
I'd still add it, because otherwise users have no clue what to do unless they go check out the new docs.

joshuacwnewton replied:
Thanks! I've re-written the message in a way that I hope make things clearer. :)

NicolasHug commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -872,13 +884,64 @@ def test_tsne_with_different_distance_metrics():
+                      random_state=random_state)
+    X_precomputed = pairwise_distances(X, metric=metric)
+
+    if metric == 'euclidean' and square_distances == 'legacy':
+        X_precomputed **= 2
```
I'm a bit confused by this condition: shouldn't this be

```py
if metric == 'euclidean' or square_distances is True
```

?

does that mean that when we have (manhattan, True) we don't expect the distances to be squared? Or are the distances squared already when calling `X_precomputed = pairwise_distances(X, metric=metric)`?

jeremiedbb replied:
with `euclidean` + `legacy`, distances are squared so to be able to compare with precomputed we need to provide a squared distance matrix.

NicolasHug replied:
But don't we expect the distances to be squared as well with `manhattan`, `True`?

jeremiedbb replied:
with True, distances are squared, whether it's a string metric (say `manhattan`) or a precomputed distance matrix (say `pairwise_distances(X, metric="manhattan")`)

NicolasHug replied:
got it, thanks

NicolasHug commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -872,13 +884,64 @@ def test_tsne_with_different_distance_metrics():
+@pytest.mark.parametrize('metric', ['euclidean', 'manhattan'])
+@pytest.mark.parametrize('square_distances', [True, 'legacy'])
+def test_tsne_square_distances_futurewarning(metric, square_distances):
+    # Make sure that a FutureWarning is only raised when a non-Euclidean
+    # metric is specified and square_metric is not set to True.
```
thank you for writing helpful test comments! these are important

joshuacwnewton replied:
Ah, I've just noticed there's a typo in the test comment you quoted! I'll fix that now. :)

NicolasHug reviewed (commented):
thanks @joshuacwnewton , some last nitpicks and one question

NicolasHug reviewed (approved):
pre-approving since the remaining comments are easy to address. Thanks @joshuacwnewton !

joshuacwnewton committed:
Fix `distance` -> `distances` in What's New

joshuacwnewton committed:
Clarify deprecation plan in What's New

joshuacwnewton committed:
Clarify deprecation plan in FutureWarning

joshuacwnewton committed:
Fix typo in test description comment

`square_metric` -> `square_distances`

joshuacwnewton commented:
Thanks much for taking the time to review, @NicolasHug! :D

jnothman commented:
Thanks you @joshuacwnewton
